cpp/src/DataStorm/SessionI.cpp

```diff
@@ -605,21 +601,22 @@ SessionI::disconnected(const Ice::ConnectionPtr& connection, exception_ptr ex)
             }
             else
             {
-                throw Ice::CloseConnectionException(__FILE__, __LINE__);
+                throw Ice::CloseConnectionException{__FILE__, __LINE__};
             }
         }
         catch (const std::exception& e)
         {
             Trace out(_traceLevels, _traceLevels->sessionCat);
-            out << _id << ": session `" << _session->ice_getIdentity() << "' disconnected:\n";
+            out << _id << ": session '" << _session->ice_getIdentity() << "' disconnected:\n";
             out << (_connection ? _connection->toString() : "<no connection>") << "\n";
             out << e.what();
         }
     }
 
+    auto self = shared_from_this();
     for (auto& t : _topics)
     {
-        runWithTopics(t.first, [&](TopicI* topic, TopicSubscriber&) { topic->detach(t.first, shared_from_this()); });
+        runWithTopics(t.first, [id = t.first, self](TopicI* topic, TopicSubscriber&) { topic->detach(id, self); });
```

Comment:
minor change, no need to call shared_from_this in the loop or capture everything by ref.

Reply:
Could you declare self in the capture block to avoid an extra copy?